src/LibDDRV.sol

```diff
@@ -0,0 +1,165 @@
+    function nlz(uint256 x) internal pure returns (uint256) {
+        if (x == 0) {
+            return 256;
+        }
+
+        uint256 n = 1;
+
+        if (x >> 128 == 0) {
+            n += 128;
+            x <<= 128;
+        }
+        if (x >> 192 == 0) {
+            n += 64;
+            x <<= 64;
+        }
+        if (x >> 224 == 0) {
+            n += 32;
+            x <<= 32;
+        }
+        if (x >> 240 == 0) {
+            n += 16;
+            x <<= 16;
+        }
+        if (x >> 248 == 0) {
+            n += 8;
+            x <<= 8;
+        }
+        if (x >> 252 == 0) {
+            n += 4;
+            x <<= 4;
+        }
+        if (x >> 254 == 0) {
+            n += 2;
+            x <<= 2;
+        }
+
+        n -= x >> 255;
+
+        return n;
+    }
+
+    function floor_ilog(uint256 x) internal pure returns (uint256) {
+        return (255 - nlz(x));
+    }
+
+    function weight(Range storage range) internal returns (uint256 w) {}
+
+    function weight(Level storage level) internal returns (uint256 w) {}
+
+    function weight(Forest storage forest) internal returns (uint256 w) {}
+
+    // Reject buckets which don't match criterion
+    function bucket_rejection() internal {}
+
+    function insert_bucket(uint256 i, Range storage range) internal {}
+
+    function delete_range(Range storage range) internal {}
+
+    // Construct a level in the forest of trees
+    function construct_level(Forest storage forest, uint256 level, bytes32 ptr, bytes32 head, bytes32 tail)
+        internal
+        returns (bytes32 ptr1, bytes32 head1, bytes32 tail1)
+    {
+        // TODO(weights, roots for a level)
+        // Qₗ₊₁ = ∅
+        assembly {
+            // Set the queue to the free pointer
+            ptr1 := mload(0x40)
+            head1 := add(ptr, 0x20)
+            // One word is reserved here to act as a header for the queue,
+            // to check if a range is already in the queue.
+            tail1 := head
+        }
+    }
+
+    // Preprocess an array of elements and their weights into a forest of trees.
+    // TODO(This presently supports natural number weights, could easily support posits)
+    function preprocess(uint256[] memory weights, Forest storage forest) external {
+        uint256 l = 1;
+        // Set up an in memory queue object
+        uint256 in_queue;
+        bytes32 ptr;
+        bytes32 head;
+        bytes32 tail;
+        // Qₗ = ∅
+        assembly {
+            // Set the queue to the free pointer
+            ptr := mload(0x40)
+            head := add(ptr, 0x20)
+            // One word is reserved here to act as a header for the queue,
+            // to check if a range is already in the queue.
+            tail := head
+        }
+        uint256 n = weights.length;
+        uint256 i;
+        uint256 j;
+        for (i = 1; i <= n; i++) {
+            j = floor_ilog(weights[i]) + i;
```

Comment:
+1, not plus `i`

![image](https://user-images.githubusercontent.com/13227294/222011623-77ffb0dc-66f4-481d-ad3d-32265efd00f8.png)